src/betterproto/streams/__init__.py

```diff
@@ -0,0 +1,82 @@
+from .. import Message, encode_varint, decode_varint
+from io import BufferedReader, BufferedWriter
+from typing import Iterable, Optional
+
+
+class CodedOutputStream:
+    stream: BufferedWriter
+
+    def __init__(self, stream: BufferedWriter) -> None:
+        self.stream = stream
+
+    def write_varint(self, value: int) -> None:
+        """Write a single varint value to the stream."""
+        self.stream.write(encode_varint(value))
+
+    def write_message(self, message: Message) -> None:
+        """Write a serialized message to the stream."""
+        self.stream.write(bytes(message))
+
+    def write_messages(self, messages: Iterable[Message]) -> None:
+        """Write multiple messages to the stream."""
+        for message in messages:
+            # Write the size of the message to delimit it and enable parsing
+            self.write_varint(len(bytes(message)))
+            self.write_message(message)
+
+
+class CodedInputStream:
+    stream: BufferedReader
+
+    def __init__(self, stream: BufferedReader) -> None:
+        self.stream = stream
+
+    def read_varint(self) -> Optional[int]:
+        """Read a single varint from the stream. Returns None if stream is at EOF."""
+        byte = self.stream.read(1)
+        if byte == b"":
+            # EOF
+            return None
+
+        # Read extended varint bytes (if varint is longer than 1 byte)
+        value = 0
+        shift = 0
+        while int.from_bytes(byte, byteorder="little") >> 7 == 1:
+            # Add current byte to the value
+            value += (int.from_bytes(byte, byteorder="little") & 0b01111111) << (7 * shift)
+
+            # Read the next byte and check it is valid
+            byte = self.stream.read(1)
+            if byte == b"":
+                # EOF is invalid, as the full varint hasn't been read yet
+                raise ValueError("Stream ended part way through a varint")
+
+            shift += 1
+
+        # Add final byte of the varint to the value
+        value += (int.from_bytes(byte, byteorder="little") & 0b01111111) << (7 * shift)
+
+        return value
+
+    def read_message(self, size: int) -> Message:
+        """Read a message from the stream of a given size.
+        Returns empty bytes object if stream is at EOF."""
+        return self.stream.read(size)
+
+    def read_messages(self, limit=0) -> Iterable[Message]:
```

Comment:
It would be nice to add a `__iter__` function that allows iterating through all the messages in the coded stream:

```py
for message in CodedInputStream(stream):
    ...
```
```

Reply:
These are now part of the `Message` class, and are `parse_from` and `parse_delimited_from`, so there isn't really a class to iterate over anymore - not sure if you wanted me to add any other features around this